qp_klp/Step.py

```diff
@@ -326,23 +328,187 @@ def _generate_prep_file(self, config, input_file_path, seqpro_path,
+
+            qiita_id = match.groups()[0]
+            uploads_folder = uploads_folders[qiita_id]
+            cmds.append(f"mv {data_dir}/*  {uploads_folder}")
+
+        return cmds
+
+    def _helper_process_operations(self):
+        MOVE_CMD = 'mv'
+        RESULTS_DIR = 'final_results'
+        ORDER_AOI = 'AOI'   # for Action, Output, Input(s)
+        ORDER_AIO = 'AIO'   # for Action, Input(s), Output
+        TAR_CMD = 'tar zcvf'
+        LOG_PREFIX = 'logs'
+        REPORT_PREFIX = 'reports'
+        PREP_PREFIX = 'prep-files'
+        CONVERT_JOB = 'ConvertJob'
+        QC_JOB = 'QCJob'
+        FASTQC_JOB = 'FastQCJob'
+        PREPFILE_JOB = 'GenPrepFileJob'
+        TAR_EXT = 'tgz'
+
+        op_meta = [(['ConvertJob/logs'], TAR_CMD,
+                    f'{LOG_PREFIX}-{CONVERT_JOB}.{TAR_EXT}', ORDER_AOI),
+
+                   (['ConvertJob/Reports', 'ConvertJob/logs'], TAR_CMD,
+                    f'{REPORT_PREFIX}-{CONVERT_JOB}.{TAR_EXT}',
+                    ORDER_AOI),
+
+                   (['QCJob/logs'], TAR_CMD,
+                    f'{LOG_PREFIX}-{QC_JOB}.{TAR_EXT}', ORDER_AOI),
+
+                   (['FastQCJob/logs'], TAR_CMD,
+                    f'{LOG_PREFIX}-{FASTQC_JOB}.{TAR_EXT}', ORDER_AOI),
+
+                   (['FastQCJob/fastqc'], TAR_CMD,
+                    f'{REPORT_PREFIX}-{FASTQC_JOB}.{TAR_EXT}', ORDER_AOI),
+
+                   (['GenPrepFileJob/logs'], TAR_CMD,
+                    f'{LOG_PREFIX}-{PREPFILE_JOB}.{TAR_EXT}', ORDER_AOI),
+
+                   (['GenPrepFileJob/PrepFiles'], TAR_CMD,
+                    f'{PREP_PREFIX}.{TAR_EXT}', ORDER_AOI),
+
+                   (['failed_samples.html', 'touched_studies.html'],
+                    MOVE_CMD, RESULTS_DIR, ORDER_AIO),
+
+                   (['FastQCJob/multiqc'], MOVE_CMD, RESULTS_DIR, ORDER_AIO)]
+
+        cmds = []
+
+        for inputs, action, output, order in op_meta:
+            confirmed_inputs = []
+            for input in inputs:
+                if exists(join(self.pipeline.output_path, input)):
+                    # it's expected that some inputs may not exist due to
+                    # different pipeline types. If one or more inputs do not
+                    # exist, do not include them in the command-line as they
+                    # may cause an error.
+                    confirmed_inputs.append(input)
+
+            # do not add the command to the list unless at least one of
+            # the inputs exists. It's okay for a command to go unprocessed.
+            if confirmed_inputs:
+                # convert to string form before using.
+                confirmed_inputs = ' '.join(confirmed_inputs)
+                if order == 'AOI':
+                    cmds.append(f'{action} {output} {confirmed_inputs}')
+                elif order == 'AIO':
+                    cmds.append(f'{action} {confirmed_inputs} {output}')
+                else:
```

Comment:
Does this level of abstraction is needed? Either way, why?

Reply:
Some actions like 'tar' require defining the output before the inputs while others like 'mv' require all inputs to be given before the output. At any rate it's been simplified some.